File: lib/prompts/templates/intro.ts
import type { Blueprint, GenerateRequest, SectionSpec } from '../../types'

export function buildPrompt(
  spec: SectionSpec,
  blueprint: Blueprint,
  req: GenerateRequest
): string {
  const { lesson_metadata } = req
  return `Write an engaging introduction for this study guide.

## Blueprint Context
- Lesson Title: ${blueprint.title}
- Essential Question: ${blueprint.essential_question}
- Introduction Hook: ${blueprint.introduction_hook}
- Core Concept: ${blueprint.core_concept}
- Subject: ${lesson_metadata.subject}
- Grade Level: ${lesson_metadata.grade_level}
- Market: ${lesson_metadata.market}

## Instructions
Write 2–3 paragraphs that:
1. Open warmly and reference the essential question to engage students
2. Briefly explain what students will learn in this lesson
3. Connect the topic to real-world relevance for Grade ${lesson_metadata.grade_level} learners in ${lesson_metadata.market}

The tone should be warm, clear, and encouraging. Never condescending. No exclamation marks.

## Output Schema
{
  "introduction": "string (2–3 paragraphs separated by \\n\\n)"
}

Section spec: ${JSON.stringify(spec)}`
}

File: backend/study-guide-agent/app/prompts/templates/intro.py
from __future__ import annotations

from app.types import Blueprint, GenerateRequest


def _intro_length_guidance(request: GenerateRequest) -> str:
    grade_level = request.lesson_metadata.grade_level

    if grade_level <= 5:
        return (
            "Keep each paragraph to 1 or 2 short sentences, and use everyday classroom "
            "words before academic phrasing."
        )
    return "Keep both paragraphs concise and focused on one main idea each."


def build_prompt(spec, blueprint: Blueprint, request: GenerateRequest) -> str:
    del spec

    intro_schema = """{
    "title": "Introduction",
    "hook": "string",
    "essential_question": "string",
    "paragraphs": [
        "string"
    ],
    "bridge_to_lesson": "string"
}"""

    learning_target_text = "\n".join(
        f"- {target.number}. {target.objective}"
        for target in blueprint.learning_targets
    )

    prompt_lines = [
        "Create the introduction section for a K-12 study guide.",
        "Use the blueprint as the single source of truth.",
        f"- Lesson title: {blueprint.title}",
        f"- Essential question: {blueprint.essential_question}",
        f"- Introduction hook: {blueprint.introduction_hook}",
        f"- Core concept: {blueprint.core_concept}",
        "- Learning targets:",
        learning_target_text,
        "Requirements:",
        "- Write a warm, student-facing opening linked directly to the essential question.",
        "- Use the hook as the first attention-grabbing idea, but rewrite it as polished study-guide prose.",
        "- Provide exactly 2 concise paragraphs that prepare students for the lesson.",
        f"- {_intro_length_guidance(request)}",
        "- Make the bridge_to_lesson explain how the opening connects to the study guide work ahead.",
        "Expected JSON schema:",
        intro_schema,
        "Return only JSON.",
    ]

    return "\n".join(prompt_lines)
